text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,991 @@
+useful for security reviews.
+
+**Question:**
+If `rand`/`Rng` is split into multiple parts, several approaches are possible;
+what to do here is very much undecided.
+
+First, "strong RNGs" designed for cryptographic usage could:
+
+1.  Use a dedicated `CryptoRng` trait, indepedant of `Rng`, with methods
+    returning a `Result` on failure
+2.  As (1), but add a wrapper struct implementing `Rng` for a `CryptoRng` which
+    panics on errors
+3.  Use the standard `Rng` trait and be designed such that the only time
+    failure is possible is during creation (`fn new() -> Result<Self>`)
+4.  Do not include any generators which may fail in `rand` (leave to dedicated
+    crypto libraries, if at all) [practically, this is equivalent to 3 aside
+    from choice of generators to include]
+
+Second, the `rand` crate could:
+
+1.  Remain a single crate
+2.  Have a sub-crate `os_rand` covering only the `OsRng` functionality (via
+    simple functions), and have `rand` depend on `os_rand` for initialisation
+3.  Have a `rand_core` sub-crate including `OsRng` functionality, the `Rng`
+    and/or `CryptoRng` trait(s), and a crypto-approved PRNG; `rand` would
+    depend on `rand_core`
+4.  Keep all generators in `rand` and move all distributions (including
+    `random()`) to a `distributions` crate (or same split but with `rng` and
+    `rand` crate names)
+5.  Split into even more crates...
+
+*Personally,* I feel that we should stick to a single `Rng` trait as above. I
+am not against splitting the implementation into multiple crates, but see
+little benefit (a split *might* facilitate review of cryptographic code).
+
+### Debug
+
+The strawman revision now specifies `pub trait Rng: Debug {...}`, i.e. requires
+all implementations to implement `Debug`. In many cases this requires only
+that implementing types are prefixed with `#[derive(Debug)]`, although in some
+cases it adds requirements on internal values.
+
+Is this useful?
+
+### Extension traits
+
+The `SeedableRng` trait should remain as is:
+
+```rust
+pub trait SeedableRng<Seed>: Rng {
+    fn reseed(&mut self, Seed);
+    fn from_seed(seed: Seed) -> Self;
+}
+```
+
+Another trait could be added to allow jump-ahead:
+
+```rust
+pub trait JumpableRng: Rng {
+    /// Return a copy of self, mutated as if `next_u32` had been called `steps`
+    /// times
+    fn jump_u32(&self, steps: usize) -> Self;
+    
+    // Also jump_u64() ? For most 64-bit generators they should be the same;
+    // for 32-bit generators, jump_u64 should jump twice as fast.
+}
+```
+
+And a trait could allow entropy injection. (Does the type `T` need to be a
+template parameter?)
+
+```rust
+pub trait InjectableRng<T>: Rng {
+    /// Add entropy to the RNG, adjusting the sequence of numbers produced.
+    fn inject(&mut self, entropy: T);
+}
+```
+
+These traits can be added in the future without breaking compatibility, however
+they may be worth discussing now.
+
+### Creation of RNGs
```

Comment:
That's the exact opposite of the correct response to the problem. People _will_ be seeding generators from other generators in the wild. They're just going to do it, even if they have to call `next_u32` manually to make it happen. Instead of providing no protection when they decide to go this route, we should anticipate it and provide as much footgun protection as we can.

* If you seed one generator from a different type of generator it seems plain that you won't accidentally end up with identical generators.
* If you seed a generator from its own type of generator, and the generator output is just the seed directly after the transition step, and then you use that output as your own seed without changing it at all, then you've just made an accidental Clone. Note that while this is a danger for many generator types it is not even possible for other generator types.

So the correct response to this situation is to encourage generator writers to provide an instance for Rand or Default or whichever trait, and have them take the footgun possibly into account. If your seed is a u64, maybe you call `next_u64` twice instead of once and wrapping_mul the values together as a way of breaking away from the other seed. Whatever works. But this way the correct code for each generator is written and reviewed once, instead of cobbled together slightly differently in every project using `rand`.

We're writing Rust, not C. Our response to danger is to do what it takes to make it safe instead.